sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala

```diff
@@ -106,11 +155,13 @@ case class Mode(
     val collationAwareBuffer = child.dataType match {
       case c: StringType if
         !CollationFactory.fetchCollation(c.collationId).supportsBinaryEquality =>
-        val collationId = c.collationId
-        val modeMap = buffer.toSeq.groupMapReduce {
-         case (k, _) => CollationFactory.getCollationKey(k.asInstanceOf[UTF8String], collationId)
+        buffer.toSeq.groupMapReduce {
+         case (k, _) => CollationFactory.getCollationKey(k.asInstanceOf[UTF8String], c.collationId)
         }(x => x)((x, y) => (x._1, x._2 + y._2)).values
-        modeMap
+      case _: ArrayType | _ : StructType if child.dataType.existsRecursively(s =>
```

Comment:
instead of manually doing `child.dataType.existsRecursively`, use `UnsafeRowUtils.isBinaryStable(child.dataType)`

Reply:
Done